src/workerd/io/compatibility-date.c++

```diff
@@ -171,6 +173,11 @@ void compileCompatibilityFlags(kj::StringPtr compatDate, capnp::List<capnp::Text
         enableByDate = parsedCompatDate >= parsedDate;
       } else if (annotation.getId() == COMPAT_ENABLE_ALL_DATES_ANNOTATION_ID) {
         enableByDate = true;
+      } else if (annotation.getId() == EXPERIMENTAl_ANNOTATION_ID) {
+        KJ_IF_MAYBE(entry, flagSet.find(enableFlagName)) {
+          isExperimental = true;
+          flagSet.erase(*entry);
+        }
```

Comment:
I didn't look closely enough at this logic before. I don't think it works. I'm a little confused how the tests could have passed.

`flagSet.find(enableFlagName)` will always return null here, because the flag was already removed from the `flagSet` when handling `COMPAT_ENABLE_FLAG_ANNOTATION_ID`.

It doesn't make sense to check the `flagSet` here, this code should just be:

```suggestion
      } else if (annotation.getId() == EXPERIMENTAl_ANNOTATION_ID) {
        isExperimental = true;
```

Of course changing it that way reveals bugs in the checks below, which are fixed by my suggestion there.

Reply:
Your existing logic here ended up never setting `isExperimental = true` at all. The reason the test passed is because your logic below was still _also_ considering flags experimental if they didn't have a compat date set, but that's logic we want to get rid of.

Please be sure to add a unit test case that verifies that a flag which has no compat date, but which is _not_ marked experimental, is allowed to be set even when experimental flags are disallowed. Perhaps you can use `obsolete14` for this purpose, since it's a flag that will definitely never have a date, but also shouldn't be considered experimental (see my other comment).

Reply:
Done, and thanks for the review!